### DISCUSSION TITLE
Modals and Popovers

### DISCUSSION THREAD

**Discussion by Arneoldenhave** [frzi/swiftui-router#52] [Q&A] [answered]
Hi! 

First of all thanks!  Great library :)

I would like to know if there is support for popovers and modals.

Kind regards,

Arne

**Comment by frzi** [comment, marked as answer]
Hello 👋 

You *can* use `.sheet()` et al with SwiftUI Router. Though it does require a bit of code. There are multiple ways to tackle this. Here's a very crude example:

```swift
struct ContentView: View {
	@EnvironmentObject private var navigator: Navigator
	@State private var showModal = false

	var body: some View {
		VStack {
			NavLink(to: "/modal") {
				Text("Go to `/modal`")
			}
			NavLink(to: "/modal/some/deeper/path") {
				Text("Go to a deeper screen in the modal")
			}
		}
		.onChange(of: navigator.path) { newPath in
			showModal = newPath.starts(with: "/modal")
		}
		.onChange(of: showModal) { newShowModal in
			if !newShowModal {
				navigator.navigate("/")
			}
		}
		.sheet(isPresented: $showModal) {
			// Recommended: wrap the modal view in a `Route` to allow for relative paths
			// *inside* the `ModalView`.
			Route("modal/*", content: ModalView())
		}
	}
}
```
Basically you check whether the current path is one that should display a sheet. Once the sheet is (manually) dismissed you programmatically navigate to another path.

Mind you the code above is not fool proof and will require some tweaking to behave more flexible. But it should give you some idea of having SwiftUI Router work with modals/popovers/whatever.

**Reply by Arneoldenhave** [reply]
Thanks for the reply! 

Unfortunately I cannot get it to work yet.


This contact list is a tabbar item as shown in your GitHub example.

`
import SwiftUI
import SwiftUIRouter

struct ContactList: View {
    
    @EnvironmentObject private var navigator: Navigator
    @EnvironmentObject private var modalState : ModalState
    
    @State
    private var showModal = false

    
    let contactModels = [
        ContactModel(name: "Harry", surname: "Potter"),
        ContactModel(name: "Hermelien", surname: "Griffoen"),
        ContactModel(name: "Ron", surname: "Wezel"),
        ContactModel(name: "Ginny", surname: "Wezel"),
        ContactModel(name: "Severus", surname: "Sneep"),
    ]
    
    var body: some View {
        List(contactModels) { contact in
            DetailRow(model: contact).onTapGesture {
                modalState.isPresented  = true
            }
        }
        .onChange(of: modalState.isPresented) { isPresented in
            self.showModal = isPresented
        }
        .onChange(of: navigator.path) { newPath in
            showModal = newPath.starts(with: "/modal")
        }
        .onChange(of: showModal) { newShowModal in
            if !newShowModal {
                navigator.navigate("/")
            }
        }
        .sheet(isPresented: $showModal) {
        // Recommended: wrap the modal view in a `Route` to allow for relative paths
        // *inside* the `ModalView`.
            Route("modal/*", content: ModalView()).background(Color.red)
        }
    }
}
`
When A user taps a detailRow a modal appears; but when it is wrapped in a Route the modal view is empty. 

When I do not warp the modal in a Route the modal appear correctly but I cannot navigate anymore .

`struct ModalView : View {
    
    @EnvironmentObject private var navigator: Navigator

    var body: some View {
        SwitchRoutes {
           Route("Green/*", content: ColorView(color: .green) )
           Route("Blue/*",  content: ColorView(color: .blue))
           Route("Red/*",   content: ColorView(color: .red))
           Route {
               ColorView(color: .red)
           }
        }
    }
}


struct ColorView: View {
    
    let color: Color

    var body: some View {
        GeometryReader { _ in
            VStack {
                Text("clorView")
                NavLink(to: "/Green") {
                    Text("greens")
                }
                NavLink(to: "/Red") {
                    Text("red")
                }
                NavLink(to: "/Blue") {
                    Text("blue")
                }
                NavLink(to: "Contacts/*") {
                    Text("Open Contancts")
                }
            }
        }.background(color)
    }
}
`


When I do not wrap the Modal in a route I get the modal inside the sheet but I cannot navigate to the colored routes after the first navigation.



One last question: I need to be able to push a View onto the root Router from a modal. Open contacts should open a new view on the stack as if a tabbar item at the root level. Would a thing like that be possible with your Router?

Thanks!
